src/libstd/sys/unix/ext/net.rs

```diff
@@ -0,0 +1,1225 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+#![unstable(feature = "unix_socket", reason = "newly added", issue = "32312")]
+
+//! Unix-specific networking functionality
+
+use libc;
+
+use prelude::v1::*;
+use ascii;
+use ffi::OsStr;
+use fmt;
+use io;
+use mem;
+use net::Shutdown;
+use os::unix::ffi::OsStrExt;
+use os::unix::io::{RawFd, AsRawFd, FromRawFd, IntoRawFd};
+use path::Path;
+use time::Duration;
+use sys::cvt;
+use sys_common::io::read_to_end_uninitialized;
+
+fn sun_path_offset() -> usize {
+    unsafe {
+        // Work with an actual instance of the type since using a null pointer is UB
+        let addr: libc::sockaddr_un = mem::uninitialized();
+        let base = &addr as *const _ as usize;
+        let path = &addr.sun_path as *const _ as usize;
+        path - base
+    }
+}
+
+struct Inner(RawFd);
+
+impl Drop for Inner {
+    fn drop(&mut self) {
+        unsafe {
+            libc::close(self.0);
+        }
+    }
+}
+
+impl Inner {
+    fn new(kind: libc::c_int) -> io::Result<Inner> {
+        unsafe { cvt(libc::socket(libc::AF_UNIX, kind, 0)).map(Inner) }
+    }
+
+    fn new_pair(kind: libc::c_int) -> io::Result<(Inner, Inner)> {
+        unsafe {
+            let mut fds = [0, 0];
+            try!(cvt(libc::socketpair(libc::AF_UNIX, kind, 0, fds.as_mut_ptr())));
+            Ok((Inner(fds[0]), Inner(fds[1])))
+        }
+    }
+
+    fn try_clone(&self) -> io::Result<Inner> {
+        unsafe { cvt(libc::dup(self.0)).map(Inner) }
```

Comment:
Similarly, this is implemented elsewhere which sets `CLOEXEC` automatically, so probably good to leverage that.